### DISCUSSION TITLE
How to programmatically add an AjaxBehavior to a UIComponent with primefaces

### DISCUSSION THREAD

**Discussion by crocimi** [primefaces/community#1611] [PrimeFaces] [answered]
I'm trying to create a CommandLink and add an AjaxBehavior programatically. The final goal is to attach (programmatically) a "dialogReturn" event.

I was able to create the CommandLink and attach the AjaxBehavior, following this post. 
[https://stackoverflow.com/questions/15808956/how-to-programmatically-add-an-ajaxbehavior-to-a-uicomponent-with-primefaces](https://stackoverflow.com/questions/15808956/how-to-programmatically-add-an-ajaxbehavior-to-a-uicomponent-with-primefaces)

The event is called and I can update some components or do some actions. BUT, if I try to open a dialog using `PrimeFaces.current().dialog().openDynamic()`, this doesn't work.

Here my code:
```
CommandLink link = (CommandLink)fc.getApplication().createComponent(CommandLink.COMPONENT_TYPE);
((CommandLink)link).setAjax(true);
link.setId("link"+"_"+row+"_"+col);
String key = this.getLinkName(table,row,col);
link.getAttributes().put(KEY, key);
link.setTabindex(String.valueOf(row));
link.setId(UIViewRoot.UNIQUE_ID_PREFIX+"_"+row+"_"+col);

ExpressionFactory ef = fc.getApplication().getExpressionFactory();
MethodExpression me = ef.createMethodExpression(fc.getELContext(),"#{myBean.linkEvent}", null, new Class<?>[]{BehaviorEvent.class});
org.primefaces.behavior.ajax.AjaxBehavior ajaxBehavior = (org.primefaces.behavior.ajax.AjaxBehavior) fc.getApplication().createBehavior(org.primefaces.behavior.ajax.AjaxBehavior.BEHAVIOR_ID);
//ajaxBehavior.setProcess("@this");
ajaxBehavior.setUpdate(reRenderId);
ajaxBehavior.addAjaxBehaviorListener(new org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl(me, me));
link.addClientBehavior(link.getDefaultEventName(), ajaxBehavior);
```

Any idea? Thx.

**Comment by melloware** [comment]
I am not saying it is not possible but certainly sounds a lot like this issue with `dialogReturn`: https://github.com/primefaces/primefaces/issues/7436

cc @christophs78

**Reply by crocimi** [reply, marked as answer]
Ok, thx. The workaround would be to have a hidden commandButton with an associated "dialogReturn" event and call it via javascript's click()

**Reply by tandraschko** [reply]
also note that PF CommandLink+Button are already ajaxified, you dont need to add f/p:ajax
